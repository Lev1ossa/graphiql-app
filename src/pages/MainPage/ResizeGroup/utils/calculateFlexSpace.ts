export function calculateFlexSpace(params: {
  fractions: number[]
  index: number
  value: number
  minmax?: [number, number][]
  sibling?: 1 | -1
}) {
  const { fractions, index, value, minmax = [], sibling = 1 } = params

  if (index < 0) return
  if (index > fractions.length - 1) return
  if (index === 0 && sibling < 0) return
  if (index === fractions.length - 1 && sibling > 0) return

  const getMin = (i: number, min = minmax?.[i]?.[0]) =>
    typeof min === 'number' ? min : 0
  const getMax = (i: number, max = minmax?.[i]?.[1]) =>
    typeof max === 'number' ? max : 1

  const sizeIndexes = fractions.map((_, i) => i)
  const pivot = sibling < 0 ? index : index + 1
  const toStart = sizeIndexes.slice(0, pivot).toReversed()
  const toEnd = sizeIndexes.slice(pivot)

  const sizes /* mut */ = [...fractions] as number[]

  const delta = value - fractions[index]

  let move = delta

  while (move) {
    const growCandidates = delta > 0 ? toStart : toEnd
    const shrinkCandidates = delta > 0 ? toEnd : toStart

    const growTarget = growCandidates.find(
      (index) => sizes[index] < getMax(index),
    )
    const shrinkTarget = shrinkCandidates.find(
      (index) => sizes[index] > getMin(index),
    )

    if (growTarget === undefined) break
    if (shrinkTarget == undefined) break

    const grow = getMax(growTarget) - sizes[growTarget]
    const shrink = sizes[shrinkTarget] - getMin(shrinkTarget)

    const step = Math.min(grow, shrink, Math.abs(move))

    sizes[growTarget] += step
    sizes[shrinkTarget] -= step

    move -= Math.sign(delta) * step
  }

  return sizes
}
