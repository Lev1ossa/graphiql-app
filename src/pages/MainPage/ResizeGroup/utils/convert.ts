export function toFr(value: number, share: number) {
  return value / share
}

export function toFrArray(values: number[]) {
  const share = values.reduce((s, v) => s + v, 0)
  return {
    fractions: values.map((v) => toFr(v, share)),
    share,
  }
}
