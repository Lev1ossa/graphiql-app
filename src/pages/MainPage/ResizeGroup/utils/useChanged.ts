import { useEffect, useRef } from 'react'

export function useChanged<T = never, V = never>(params: {
  value: V
  transform?: (v: V) => T extends never ? V : T
}) {
  const { value, transform = (v) => v } = params
  const prevValueRef = useRef(value)
  useEffect(() => {
    prevValueRef.current = value
  })
  return (() => {
    const prev = prevValueRef.current
    const curr = value
    return transform(prev) !== transform(curr)
  })()
}
