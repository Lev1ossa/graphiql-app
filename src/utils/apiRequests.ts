import { buildClientSchema, getIntrospectionQuery, printSchema } from 'graphql'

export const getApiIntrospectionSchema = () => {
  return fetch('https://countries.trevorblades.com/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ query: getIntrospectionQuery() }),
  })
    .then((res) => res.json())
    .then((result) => {
      const clientSchema = buildClientSchema(result.data)
      return printSchema(clientSchema)
    })
}

// TODO change api url from default to user input
export const getApiResponse = (
  schema: string,
  headers: Record<string, string>,
  variables?: Record<string, string>,
) => {
  return fetch('https://countries.trevorblades.com/graphql', {
    method: 'POST',
    headers,
    body: JSON.stringify({
      query: schema,
      variables,
    }),
  })
    .then((res) => res.json())
    .then((result) => {
      return result.data
    })
}

// TODO remove all examples
// const ExampleSchema = `
//   query countries($filter: CountryFilterInput) {
//     countries(filter: $filter) {
//       __typename
//       code
//       name
//       native
//       phone
//       capital
//       currency
//       emoji
//       emojiU
//     }
//   }
// `;

// const exampleHeaders =  {
//   'Content-Type': 'application/json',
// };
