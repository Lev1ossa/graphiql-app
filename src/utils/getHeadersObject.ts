import type { Header } from '@/store/store'

export function getHeadersObject(headers: Header[]) {
  const cleanedHeaders = {} as Record<string, string>
  cleanedHeaders['Content-type'] = 'application/json'
  headers.forEach((header) => {
    if (header.checked) {
      cleanedHeaders[header.headerKey] = header.headerVal
    }
  })
  return cleanedHeaders
}
