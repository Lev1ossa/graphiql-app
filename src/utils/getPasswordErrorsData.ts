export const getPasswordErrorsData = (
  passwordErrors: string[] | null,
  styles: CSSModuleClasses,
) => {
  if (passwordErrors && passwordErrors.length === 5) {
    return { className: `${styles.indicator}`, text: '' }
  }
  if (passwordErrors && passwordErrors.length === 4) {
    return {
      className: `${styles.indicator} ${styles.password_very_weak}`,
      text: `Very weak`,
    }
  }
  if (passwordErrors && passwordErrors.length === 3) {
    return {
      className: `${styles.indicator} ${styles.password_weak}`,
      text: `Weak`,
    }
  }
  if (passwordErrors && passwordErrors.length === 2) {
    return {
      className: `${styles.indicator} ${styles.password_normal}`,
      text: `So-so`,
    }
  }
  if (passwordErrors && passwordErrors.length === 1) {
    return {
      className: `${styles.indicator} ${styles.password_good}`,
      text: `Good`,
    }
  }
  return {
    className: `${styles.indicator} ${styles.password_strong}`,
    text: `Great!`,
  }
}
