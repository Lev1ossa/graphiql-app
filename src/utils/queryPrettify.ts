export const queryPrettify = (query: string) => {
  const indent = (s = '', tab = '  ') => {
    let out = ''
    for (let lvl = 0, i = 0; i < s.length; i++) {
      if (s[i] === '{') lvl += 1
      if (s[i] === '}') lvl -= 1
      if (s[i - 1] === '\n') out += tab.repeat(lvl)
      out += s[i]
    }
    return out
  }
  return indent(
    query
      .replaceAll(/\n{2,}/g, '')
      .split('\n')
      .map((line) => line.trim())
      .map((line) =>
        line.startsWith('#')
          ? line
          : line
              .replaceAll(/[\n]{2,}/g, '\n')
              .replaceAll(/[ ]{2,}/g, ' ')
              .replaceAll(/[ ]*[(]/g, '(')
              .replaceAll(/[ ]*[)]/g, ')')
              .replaceAll(/:[ ]*/g, ': ')
              .replaceAll(/[ ]*=[ ]*/g, ' = '),
      )
      .join('\n')
      .replaceAll(/[ \n]*[{][ \n]*/g, ' {\n')
      .replaceAll(/[ \n]*[}]/g, '\n}'),
  ).trim()
}
