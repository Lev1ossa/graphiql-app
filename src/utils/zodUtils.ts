import { z } from 'zod'

export const passwordSchema = z.object({
  password: z
    .string()
    .regex(/[A-Z]/g, { message: 'one uppercase letter' })
    .regex(/[a-z]/g, { message: 'one lowercase letter' })
    .regex(/\d/, { message: 'one digit' })
    .regex(/[!@#$%^&*(),.?":{}|<>]/, { message: 'one special character' })
    .min(8, { message: 'eight characters or more' }),
})

export const signinSchema = passwordSchema.extend({
  email: z.string().min(1, { message: 'Email is required' }).email(),
})

export const signupSchema = signinSchema
  .extend({
    confirmPassword: z
      .string()
      .min(1, { message: 'Confirm Password is required' }),
  })
  .refine((data) => data.password === data.confirmPassword, {
    path: ['confirmPassword'],
    message: "Password don't match",
  })
