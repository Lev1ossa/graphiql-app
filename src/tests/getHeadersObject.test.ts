import { getHeadersObject } from '@/utils/getHeadersObject'
import { expect, test } from 'vitest'

test('getHeadersObject filters data', () => {
  expect(
    getHeadersObject([
      { id: 1, checked: true, headerKey: 'key1', headerVal: 'val1' },
      { id: 2, checked: true, headerKey: 'key2', headerVal: 'val2' },
      { id: 3, checked: true, headerKey: 'key3', headerVal: 'val3' },
      { id: 4, checked: true, headerKey: 'key4', headerVal: 'val4' },
    ]),
  ).toMatchSnapshot()
})
