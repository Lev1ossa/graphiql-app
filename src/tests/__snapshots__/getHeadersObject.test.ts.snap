// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getHeadersObject filters data 1`] = `
{
  "Content-type": "application/json",
  "key1": "val1",
  "key2": "val2",
  "key3": "val3",
  "key4": "val4",
}
`;
